migrator/src/main/scala/com/scylladb/migrator/scylla/ScyllaValidator.scala

```diff
@@ -0,0 +1,133 @@
+        sourceTableDef.regularColumns.flatMap { colDef =>
+          val alias = config.renamesMap(colDef.columnName)
+
+          if (sourceSettings.preserveTimestamps)
+            List(
+              ColumnName(colDef.columnName, Some(alias)),
+              WriteTime(colDef.columnName, Some(alias + "_writetime")),
+              TTL(colDef.columnName, Some(alias + "_ttl"))
+            )
+          else List(ColumnName(colDef.columnName, Some(alias)))
+        }
+
+      val primaryKeyProjection =
+        (sourceTableDef.partitionKey ++ sourceTableDef.clusteringColumns)
+          .map(colDef => ColumnName(colDef.columnName, config.renamesMap.get(colDef.columnName)))
+
+      val consistencyLevel = sourceSettings.consistencyLevel match {
+        case "LOCAL_QUORUM" => ConsistencyLevel.LOCAL_QUORUM
+        case "QUORUM"       => ConsistencyLevel.QUORUM
+        case "LOCAL_ONE"    => ConsistencyLevel.LOCAL_ONE
+        case "ONE"          => ConsistencyLevel.ONE
+        case _              => ConsistencyLevel.LOCAL_QUORUM
+      }
+      if (consistencyLevel.toString == sourceSettings.consistencyLevel) {
+        log.info(
+          s"Using consistencyLevel [${consistencyLevel}] for VALIDATOR SOURCE based on validator source config [${sourceSettings.consistencyLevel}]")
+      } else {
+        log.info(
+          s"Using DEFAULT consistencyLevel [${consistencyLevel}] for VALIDATOR SOURCE based on unrecognized validator source config [${sourceSettings.consistencyLevel}]")
+      }
+
+      spark.sparkContext
+        .cassandraTable(sourceSettings.keyspace, sourceSettings.table)
+        .withConnector(sourceConnector)
+        .withReadConf(
+          ReadConf
+            .fromSparkConf(spark.sparkContext.getConf)
+            .copy(
+              splitCount       = sourceSettings.splitCount,
+              fetchSizeInRows  = sourceSettings.fetchSize,
+              consistencyLevel = consistencyLevel
+            )
+        )
+        .select(primaryKeyProjection ++ regularColumnsProjection: _*)
+    }
+
+    val joined = {
+      val regularColumnsProjection =
+        sourceTableDef.regularColumns.flatMap { colDef =>
+          val renamedColName = config.renamesMap(colDef.columnName)
+
+          if (sourceSettings.preserveTimestamps)
+            List(
+              ColumnName(renamedColName),
+              WriteTime(renamedColName, Some(renamedColName + "_writetime")),
+              TTL(renamedColName, Some(renamedColName + "_ttl"))
+            )
+          else List(ColumnName(renamedColName))
+        }
+
+      val primaryKeyProjection =
+        (sourceTableDef.partitionKey ++ sourceTableDef.clusteringColumns)
+          .map(colDef => ColumnName(config.renamesMap(colDef.columnName)))
+
+      val joinKey = (sourceTableDef.partitionKey ++ sourceTableDef.clusteringColumns)
+        .map(colDef => ColumnName(config.renamesMap(colDef.columnName)))
+
+      source
+        .leftJoinWithCassandraTable(
```

Comment:
when reviewing it I realized there is likely a nice optimization that instead of left join one can use something like
https://github.com/raghu-nandan-bs/cassandra-data-migrator/pull/1
- long story short - you query on both sides using same token range(s), so instead single selects from target you get a batch reply, but let me make a new GH on this idea